command/agent/config/config.go

```diff
@@ -350,6 +351,140 @@ func (c *Config) ValidateConfig() error {
+	}
+
+	if len(c.Templates) > 0 {
+		return fmt.Errorf("'template' cannot be specified with 'env_template' entries")
+	}
+
+	if c.Exec == nil || len(c.Exec.Command) == 0 {
+		return fmt.Errorf("'exec' requires non-empty 'command' field")
+	}
+
+	uniqueKeys := make(map[string]struct{})
+
+	for _, template := range c.EnvTemplates {
+		// only two fields are currently required (name & contents)
+		if template.MapToEnvironmentVariable == nil {
+			return fmt.Errorf("env_template: an environment variable name is required")
+		}
+
+		key := *template.MapToEnvironmentVariable
+
+		if _, exists := uniqueKeys[key]; exists {
+			return fmt.Errorf("env_template: duplicate environment variable name: %q", key)
+		}
+
+		uniqueKeys[key] = struct{}{}
+
+		if template.Contents == nil {
+			return fmt.Errorf("env_template[%s]: 'contents' is required", key)
+		}
+
+		if template.Backup != nil {
+			return fmt.Errorf("env_template[%s]: 'backup' is not allowed", key)
+		}
+
+		if template.Command != nil {
+			return fmt.Errorf("env_template[%s]: 'command' is not allowed", key)
+		}
+
+		if template.CommandTimeout != nil {
+			return fmt.Errorf("env_template[%s]: 'command_timeout' is not allowed", key)
+		}
+
+		if template.CreateDestDirs != nil {
+			return fmt.Errorf("env_template[%s]: 'create_dest_dirs' is not allowed", key)
+		}
+
+		if template.Destination != nil {
+			return fmt.Errorf("env_template[%s]: 'destination' is not allowed", key)
+		}
+
+		if template.ErrFatal != nil {
+			return fmt.Errorf("env_template[%s]: 'error_fatal' is not allowed", key)
+		}
+
+		if template.Exec != nil {
+			return fmt.Errorf("env_template[%s]: 'exec' is not allowed", key)
+		}
+
+		if template.Perms != nil {
+			return fmt.Errorf("env_template[%s]: 'perms' is not allowed", key)
+		}
+
+		if template.User != nil {
+			return fmt.Errorf("env_template[%s]: 'user' is not allowed", key)
+		}
+
+		if template.Uid != nil {
+			return fmt.Errorf("env_template[%s]: 'uid' is not allowed", key)
+		}
+
+		if template.Group != nil {
+			return fmt.Errorf("env_template[%s]: 'group' is not allowed", key)
+		}
+
+		if template.Gid != nil {
+			return fmt.Errorf("env_template[%s]: 'gid' is not allowed", key)
+		}
+
+		if template.Source != nil {
+			return fmt.Errorf("env_template[%s]: 'source' is not allowed", key)
+		}
+
+		if template.Wait != nil {
+			return fmt.Errorf("env_template[%s]: 'wait' is not allowed", key)
+		}
+
+		if template.LeftDelim != nil {
+			return fmt.Errorf("env_template[%s]: 'left_delimiter' is not allowed", key)
+		}
+
+		if template.RightDelim != nil {
+			return fmt.Errorf("env_template[%s]: 'right_delimiter' is not allowed", key)
+		}
+
+		if template.ExtFuncMap != nil {
+			return fmt.Errorf("env_template[%s]: 'right_delimiter' is not allowed", key)
+		}
+
+		if len(template.FunctionDenylist) != 0 {
+			return fmt.Errorf("env_template[%s]: 'function_denylist' is not allowed", key)
```

Comment:
Why should this be disallowed for env templates?

Reply:
I was being overzealous marking everything as disallowed by default. What is the typical usage of this field?

Reply:
This is what's said in the consul-template docs (https://github.com/hashicorp/consul-template/blob/main/docs/configuration.md):

```
  # These are functions that are not permitted in the template. If a template
  # includes one of these functions, it will exit with an error.
  function_denylist = []
```

I don't see why we shouldn't support this, if possible. I think it makes sense to -- as much as possible -- only disallow the fields that don't apply to this use case.

On this note, I think `error_fatal` can be allowed, too:
```
  # This controls whether an error within the template will cause
  # consul-template to immediately exit.
  error_fatal = true
```

Reply:
Thanks for the explanation, yes it makes sense to support this.